### ISSUE TITLE
Clarification on ExportedProgram.constants and its relationship to parameters, buffers, and lifted constants

### ISSUE THREAD

**Issue by mhs4670go** [Samsung/TICO#13]
## What

While using `torch.export` to convert a model into an `ExportedProgram`, I encountered multiple ways that tensor values are handled and categorized. These include:

- ExportedProgram.inputs_to_parameters
- ExportedProgram.inputs_to_buffers
- ExportedProgram.inputs_to_lifted_tensor_constants
- ExportedProgram.constants

At first glance, these seem to overlap in purpose — after all, parameters, buffers, and constants are all constant values during inference. However, they are separated for a reason, and their handling in `ExportedProgram` varies significantly. This issue aims to clarify their differences and explore some edge cases, especially around `.constants`.

Note that I'm opening this issue because I haven't seen able to find any official documentation that clarifies the differences between these categories.

### 1. Parameters

- Tensors wrapped with `nn.Parameter`.
- Trainable (i.e., require gradients).
- Included in the model’s state_dict.
- Appears in `ExportedProgram.inputs_to_parameters`.
- Must be provided again when re-executing the exported program, because:
  - `ExportedProgram` is **stateless** — it does not store parameters internally.
  - Parameters are treated as **explicit inputs** to make the program modular, portable, and backend/compiler-friendly.

### 2. Buffers

- Registered using `register_buffer()`.
- Not trainable, but part of the model state (e.g., BatchNorm’s `running_mean`).
- Included in `state_dict`.
- Typically appear in `inputs_to_buffers`.
- However, sometimes they appear in `.constants` instead, if Pytorch decides they don’t need to be passed as input or stored in the state.
  - e.g. if used only in a read-only way and not explicitly lifted as inputs. 

### 3. Lifted Tensor Constants

- Originally inline tensors in the code (e.g., `torch.tensor([1.0, 2.0])`).
- Automatically lifted as *inputs* during export when `lift_constant=True`.
- Appear in `inputs_to_lifted_tensor_constants`.
- Useful for decoupling hardcoded values from the graph for later replacement or analysis.

### 4. Constants (`.constants`)

- A dictionary inside `ExportedProgram` holding **graph-level embedded constants**.
- Used for any value considered immutable and not passed in as input.
- Includes:
  - Hardcoded inline constants (e.g., masks, fixed scalars),
  - Parameters or buffers that were *not lifted as inputs*,
  - Possibly even lifted constants that weren’t explicitly marked as such.

### Example Case: Buffer in .constants

```python
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer("scale", torch.tensor([0.1, 0.2]))

    def forward(self, x):
        return x * self.scale
```

Depending on export behavior, scale may appear:
- In inputs_to_buffers (preferred if buffer is to be passed in)
- Or directly inside .constants if treated as an immutable value during tracing

Sometimes, this creates confusion: the same buffer may or may not appear in .constants depending on its usage or internal heuristics.

## Comparison Table

| Category | Trainable | Saved in state_dict | Passed as input? | Stored in .constants | Typical Source |
| --- | :---: | :---: | :---: | :---: | --- |
| **Parameter** | Yes | Yes | Yes | No | nn.Parameter |
| **Buffer** | No | Yes | Yes or No | Sometimes | register_buffer() |
| **Lifted Tensor Constant** | No | No | Yes | No | Inline tensor lifted as input |
| **Constant** | No | No | No | Yes | Hardcoded tensors, unused buffers/params |

## Why the Distinction Matters

These distinctions affect how the model is:
- Serialized: `state_dict` restoration depends on identifying parameters and buffers.
- Reconstructed: Inputs must be mapped back correctly to maintain graph semantics.
  - For `ExportedProgram`, any value lifted as an input (parameter, buffer, or constant) must be explicitly passed again on execution.
- Optimized/Compiled: Downstream tools may fuse or re-order operations based on whether values are constant, trainable, or mutable.
- Controlled by users: For advanced use cases (e.g., compiler integration, quantization, cross-device deployment), users may want explicit control over how values are classified.

## Questions and Tentative Answers

**Q1.** What is the rule or heuristic that decides whether a buffer ends up in `inputs_to_buffers` vs `.constants`?

**A1.**

It depends on whether the buffer is referenced as a model state or just used as a read-only value during tracing. If the buffer is only read and not mutated, and isn’t marked to be reused externally, it may be lifted as a constant. This decision is affected by usage, model structure, and export-time analysis.

**Q2.** Can users explicitly control this behavior?

**A2.**

There is currently limited user control. While you can rewrite the model to lift values as inputs manually, there’s no direct API to force buffers into `inputs_to_buffers` vs `.constants`. Some control may be achieved by manipulating model structure or export options, but it’s not fully transparent.

**Q3.** Is `.constants` a stable API or an internal detail?

**A3.**

`.constants` is part of the public `ExportedProgram` API and is safe to use for retrieving embedded values. However, the naming of keys (e.g., 'constant_0') is not guaranteed to be stable across PyTorch versions or export runs.

**Q4.** Are there guarantees about how `.constants` keys map to nodes in the IR?

**A4.**

No strict guarantees. Keys like 'constant_0' are referenced by name in the FX graph, but users should not rely on naming stability. To map a constant to its use, inspect the graph and match the key manually.

## Conclusion

The distinction between parameters, buffers, lifted constants, and embedded constants is subtle but important. Better documentation and tooling to help inspect and control value classification during export would be very helpful — especially for compiler developers, deployment tooling, and advanced use cases.

Thanks for your consideration and any clarification you can provide!

**Comment by seanshpark** [comment]
maybe https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/docs/serialization.md will help a little
